node/lease_controller_v1.go

```diff
@@ -0,0 +1,341 @@
+package node
+
+/*
+Copyright 2018 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+import (
+	"context"
+	"errors"
+	"fmt"
+	"time"
+
+	"github.com/virtual-kubelet/virtual-kubelet/log"
+	"github.com/virtual-kubelet/virtual-kubelet/trace"
+	coordinationv1 "k8s.io/api/coordination/v1"
+	corev1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/wait"
+	coordclientset "k8s.io/client-go/kubernetes/typed/coordination/v1"
+	"k8s.io/utils/clock"
+	"k8s.io/utils/pointer"
+)
+
+const (
+	// renewIntervalFraction is the fraction of lease duration to renew the lease
+	renewIntervalFraction = 0.25
+	// maxUpdateRetries is the number of immediate, successive retries the Kubelet will attempt
+	// when renewing the lease before it waits for the renewal interval before trying again,
+	// similar to what we do for node status retries
+	maxUpdateRetries = 5
+	// maxBackoff is the maximum sleep time during backoff (e.g. in backoffEnsureLease)
+	maxBackoff = 7 * time.Second
+)
+
+// leaseController is a v1 lease controller and responsible for maintaining a server-side lease as long as the node
+// is healthy
+type leaseController struct {
+	leaseClient          coordclientset.LeaseInterface
+	leaseDurationSeconds int32
+	renewInterval        time.Duration
+	clock                clock.Clock
+	pingStatusFunc       func(context.Context) (*pingResult, error)
+	getNodeFunc          func(context.Context) (*corev1.Node, error)
+
+	// latestLease is the latest node lease which Kubelet updated or created
+	latestLease *coordinationv1.Lease
+}
+
+// newLeaseController constructs and returns a v1 lease controller where the renewalinterval is 0.25 * the lease interval.
+func newLeaseController(
```

Comment:
How come we have two private functions for this? I see one is with interval and one passes an interval down, but these aren't used anywhere.